src/sqlfluff/core/rules/std/L043.py

```diff
@@ -0,0 +1,262 @@
+            # same as an external table. Return the "table" part as a string.
+            table_expr = segment.get_child("main_table_expression")
+            if table_expr:
+                return table_expr.raw
+        return buff
+
+    def __init__(self, select_statement, dialect):
+        self.select_statement = select_statement
+        self.dialect = dialect
+
+    @cached_property
+    def select_info(self):
+        """Returns SelectStatementColumnsAndTables on the SELECT."""
+        result = L020.Rule_L020.get_select_statement_info(
+            self.select_statement, self.dialect, early_exit=False
+        )
+        return result
+
+    def find_alias(self, table: str) -> Optional[AliasInfo]:
+        """Find corresponding table_aliases entry (if any) matching "table"."""
+        alias_info = [
+            t
+            for t in self.select_info.table_aliases
+            if t.aliased and t.ref_str == table
+        ]
+        assert len(alias_info) <= 1
+        return alias_info[0] if alias_info else None
+
+    def get_wildcard_info(self) -> List[WildcardInfo]:
+        """Find wildcard (*) targets in the SELECT."""
+        buff = []
+        for seg in self.select_info.select_targets:
+            if seg.get_child("wildcard_expression"):
+                if "." in seg.raw:
+                    # The wildcard specifies a target table.
+                    table = seg.raw.rsplit(".", 1)[0]
+                    buff.append(WildcardInfo(seg, [table]))
+                else:
+                    # The wildcard is unqualified (i.e. does not specify a
+                    # table). This means to include all columns from all the
+                    # tables in the query.
+                    buff.append(
+                        WildcardInfo(
+                            seg,
+                            [
+                                alias_info.ref_str
+                                for alias_info in self.select_info.table_aliases
+                            ],
+                        )
+                    )
+        return buff
+
+    @staticmethod
+    def _get_name_if_cte(
+        select_statement: BaseSegment, ancestor_segment: BaseSegment
+    ) -> Optional[str]:
+        """Return name if CTE. If top-level, return None."""
+        cte = None
+        path_to = ancestor_segment.path_to(select_statement)
+        for seg in path_to:
+            if seg.is_type("common_table_expression"):
+                cte = seg
+                break
+        select_name = cte.segments[0].raw if cte else None
+        return select_name
+
+
+class Rule_L043(BaseCrawler):
+    """Query produces an unknown number of result columns.
+
+    | **Anti-pattern**
+    | Querying all columns using `*` produces a query result where the number
+    | or ordering of columns may vary due to schema changes in upstream data
+    | sources. This should be avoided because it is prone to breakage in
+    | production.
```

Comment:
Is it not the case that using `*` leaves you less prone to breakages, by the nature of it being a wildcard? My motivation for using a rule like this is being explicit about the columns produced by a query, and that schema changes are caught when you might otherwise be unaware.

Reply:
It's a good point. I wrote this text without thinking about it as deeply as I could. I'll update to address your point and try to do some internet reading to get other perspectives too.

I think the risk of breakage I mentioned is more to downstream consumers -- if they only _expect_ certain columns, but you give them more, who knows how they'll behave? It's also just untidy. For example, if you are publishing a table to a wide audience, you might "leak" additional data to them unintentionally. Real-world example: At my employer, the BigQuery data warehouse is assembled from hundreds of production MySQL servers, and some tables actually include MySQL host names. If we accidentally leaked that field to a contractor employee or the outside world, that'd be a security risk.

Reply:
In reading, I found one case where `SELECT *` could break _your own code`: `UNION` and `DIFFERENCE` require the inputs to have the same number of columns in the same order. If new columns are added to one input and not the other, I think the query will fail at runtime. Also, `SELECT *` passes through the sequence of the upstream columns, so if that changes (even though no columns were added or deleted), bad things may happen: runtime failure or wacky results (note that SQL does not check that the column *names* match, so you could be mixing completely different fields).

I decided to create a separate issue (#810) about `UNION` and `DIFFERENCE`. It's very much the same kind of concern, but it's a change in scope, and I think the rule provided by the current PR is a useful new feature by itself.